### PR TITLE
451001-YuryRybak

### PR DESCRIPTION
**Pull request by YuryRybak**

### CODE REVIEW

File: Mapper/DtoMapper.Tests/MapperTests.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            SameType = SourceForCacheTest.SameType,
+            Convertible = SourceForCacheTest.Convertible
+        };
+
+        [Test]
+        public void Map_NullPassed_ExceptionThrown()
+        {
+            var mapper = new Mapper();
+
+            Assert.Throws<ArgumentNullException>(() => mapper.Map<object, object>(null));
```
Comment (RamanBut-Husaim) [inline Mapper/DtoMapper.Tests/MapperTests.cs]: @YuryRybak It is better to declare a lambda called `act` and call it inside Assert.

File: Mapper/DtoMapper/CacheProvider.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            public int GetHashCode(TypesPair obj)
+            {
+                if (obj == null) throw new ArgumentNullException(nameof(obj));
+
+                var sourceHash = obj.Source?.GetHashCode() ?? 0;
+                var destinationHash = obj.Destination?.GetHashCode() ?? 0;
+
+                unchecked
+                {
+                    return 2147483647 * sourceHash ^ destinationHash;
```
Comment (RamanBut-Husaim) [inline Mapper/DtoMapper/CacheProvider.cs]: @YuryRybak Why have you used such algorithm?

File: Mapper/DtoMapper/CacheProvider.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace DtoMapper
+{
+    internal class CacheProvider
```
Comment (RamanBut-Husaim) [inline Mapper/DtoMapper/CacheProvider.cs]: @YuryRybak In general it is better to implement cash as a decorator for your lambda generation functionality.

File: Mapper/DtoMapper/ConditionsValidator.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (destination == null) throw new ArgumentNullException(nameof(destination));
+
+            var sourceTypeCode = Type.GetTypeCode(source);
+            var destTypeCode = Type.GetTypeCode(destination);
+
+            switch (sourceTypeCode)
+            {
+                case TypeCode.Char:
+                    return new[]
```
Comment (RamanBut-Husaim) [inline Mapper/DtoMapper/ConditionsValidator.cs]: @YuryRybak These arrays will be created each time you call this method...

**Comment by YuryRybak** [comment]
Fixed. Please, take a look.

**Comment by YuryRybak** [comment]
Fixed.
@RamanBut-Husaim please, take a look.

**Pull request merged**